packages/contracts/src/strategies/AaveSupplyStrategy.sol

```diff
@@ -0,0 +1,232 @@
+  }
+
+  function __AaveSupplyStrategy_init_unchained(
+    IAavePool _pool,
+    uint256 _millisecondsPerBlock,
+    uint256 _aaveVersion
+  ) internal onlyInitializing {
+    pool = IAavePool(_pool);
+    millisecondsPerBlock = _millisecondsPerBlock;
+
+    aaveVersion = _aaveVersion;
+    address aTokenAddress = address(0);
+
+    address assetAddress = address(asset);
+    address poolAddress = address(_pool);
+
+    if (_aaveVersion == 3) {
+      aTokenAddress = IAaveV3Pool(poolAddress).getReserveData(assetAddress).aTokenAddress;
+    } else {
+      aTokenAddress = IAaveV2Pool(poolAddress).getReserveData(assetAddress).aTokenAddress;
+    }
+
+    if (aTokenAddress == address(0)) {
+      revert UnsupportedAssetByPool();
+    }
+    aToken = IERC20Upgradeable(aTokenAddress);
+
+    // In each case, the decimal numbers of aToken must be equal to the decimal numbers of the asset
+    if (IERC20MetadataUpgradeable(aTokenAddress).decimals() != IERC20MetadataUpgradeable(address(asset)).decimals()) {
+      revert IncompatibleDecimals();
+    }
+
+    // Give full asset allowance to the pool
+    approveTokenMax(address(asset), address(pool));
+  }
+
+  /// @notice Estimates the total amount of strategy funds (including those invested in the base protocol).
+  function estimatedTotalAssets() public view override returns (uint256) {
+    return _freeAssets() + aToken.balanceOf(address(this));
+  }
+
+  /// @notice The main function of the strategy.
+  /// By calling this function, the strategy must realize (take out) the possible profits from the underlying protocol.
+  function _harvest(
+    uint256 outstandingDebt
+  ) internal override returns (uint256 profit, uint256 loss, uint256 debtPayment) {
+    profit = 0;
+    loss = 0;
+
+    // No positions to harvest, allocate available funds to pay the debt (if any)
+    if (aToken.balanceOf(address(this)) == 0) {
+      debtPayment = MathUpgradeable.min(_freeAssets(), outstandingDebt);
+      return (profit, loss, debtPayment);
+    }
+
+    uint256 deposits = aToken.balanceOf(address(this));
+    uint256 assetBalance = _freeAssets();
+    uint256 balance = deposits + assetBalance;
+
+    uint256 debt = lender.currentDebt();
+
+    if (balance <= debt) {
+      loss = debt - balance;
+      debtPayment = MathUpgradeable.min(assetBalance, outstandingDebt);
+      return (profit, loss, debtPayment);
+    }
+
+    profit = balance - debt;
+    if (assetBalance < profit) {
+      _withdrawFromProtocol(profit - assetBalance);
+      assetBalance = _freeAssets();
+      debtPayment = MathUpgradeable.min(assetBalance, outstandingDebt);
+      profit = assetBalance - debtPayment;
+      return (profit, loss, debtPayment);
+    }
+
+    if (assetBalance > profit + outstandingDebt) {
+      debtPayment = outstandingDebt;
+      return (profit, loss, debtPayment);
+    }
+
+    debtPayment = assetBalance - profit;
+    return (profit, loss, debtPayment);
+  }
```

Comment:
For now it okay, but in future will be good to resuse this logic across strategies. To avoid repetable bugs